File: source/js/mentions.js
/* mentions.js -- handle post mentions/reactions */

document.addEventListener("DOMContentLoaded", () => {
    async function getMentions() {
        try {
            const path = window.location.pathname;
            const response = await fetch(`/webmentions/get?url=${encodeURIComponent(path)}`);
            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }
            const data = await response.json();
            const mentions = data.mentions;

            if (Array.isArray(mentions) && mentions.length > 0) {
                const reactionsDiv = document.createElement("div");
                reactionsDiv.id = "reaction-container";

                const heading = document.createElement("h3");
                heading.textContent = "Reactions";
                reactionsDiv.appendChild(heading);

                const ul = document.createElement("ul");
                mentions.forEach(mention => {
                    const li = document.createElement("li");
                    const profileLink = document.createElement("a");
                    profileLink.href = mention.hcard.homepage;
                    const img = document.createElement("img");
                    img.src = mention.hcard.avatar || "/assets/img/no_avatar.png";
                    img.className = "mention_avatar";
                    img.alt = `${mention.hcard.name}’s avatar`;
                    img.onerror = () => {
                        img.src = "/assets/img/no_avatar.png";
                    };
                    profileLink.appendChild(img);
                    profileLink.innerHTML += ` ${mention.hcard.name}`;

                    li.appendChild(profileLink);

                    const mentionTypes = {
                        "bookmark": "bookmarked",
                        "like": "liked",
                        "reply": "replied to",
                        "repost": "reposted"
                    };
                    let action = mentionTypes[mention.type] || "mentioned";
                    if (mention.source_url.includes("likes/")) {
                        action = "liked";
                    }

                    const actionLink = document.createElement("a");
                    actionLink.href = mention.source_url;
                    actionLink.textContent = action;
                    li.appendChild(document.createTextNode(" "));
                    li.appendChild(actionLink);
                    li.appendChild(document.createTextNode(" this post"));
                    ul.appendChild(li);
                });
                reactionsDiv.appendChild(ul);

                const postArticle = document.querySelector("article:last-of-type");
                postArticle.insertAdjacentElement("afterend", reactionsDiv);
            }
        } catch (error) {
            console.error("Error fetching reactions:", error);
        }
    }
    getMentions();
});

File: utilities/mentions.py
import httpx
import json


from mentions.models import Webmention


def get_mentions(content_id):
    print(f"Getting mentions for ID: {content_id}")
    mentions = (
        Webmention.objects.filter(object_id=content_id, approved=1)
        .select_related("hcard")
        .order_by("-created_at")
    )

    for mention in mentions:
        try:
            with httpx.Client(timeout=5.0) as client:
                response = client.head(mention.hcard.avatar)
                if response.status_code == 200:
                    mention.hcard.avatar = mention.hcard.avatar
                else:
                    mention.hcard.avatar = None
        except httpx.RequestError:
            mention.hcard.avatar = None

        hcard_meta = json.loads(mention.hcard.json)

        if hcard_meta and not mention.hcard.homepage:
            mention.hcard.homepage = hcard_meta.url if hcard_meta.url != "/" else None

        if "likes/" in mention.source_url or mention.post_type == "like":
            mention.action = "liked"
        elif mention.post_type == "bookmark":
            mention.action = "bookmarked"
        elif mention.post_type == "reply":
            mention.action = "replied to"
        elif mention.post_type == "repost":
            mention.action = "reposted"
        else:
            mention.action = "mentioned"

    return mentions
